### PR TITLE
Add support On Demand Resources

### LINKED ISSUE #750: Setting On-Demand-Resource Tags
Is there way to setting on-demand-resource tags?

### PR DESCRIPTION
**Pull request by sipao**
Resolves #750

```
sources:
  - path: example.mp4
    resourceTags: ["example", "movie"]
```

If the file exists directly under `Foo` and `Foo` is specified in the sources, need to use excluded.

```
sources:
  - path: Foo
    excludes:
      - "example.mp4"
  - path: Foo/example.mp4
    resourceTags: ["example", "movie"]
```

### CODE REVIEW

File: Docs/ProjectSpec.md [hunk 1]
```diff
@@ -342,6 +342,7 @@ A source can be provided via a string (the path) or an object of the form:
 	- `private`
 	- `project`
 - [ ] **attributes**: **[String]** - Additional settings attributes that will be applied to any build files.
+- [ ] **resourceTags**: **[String]** - Additional On Demand Resource Tags that will be applied to any build files. this also add to project attribute's  knownAssetTags
```
Comment (yonaskolb) [inline Docs/ProjectSpec.md]:
```suggestion
- [ ] **resourceTags**: **[String]** - On Demand Resource Tags that will be applied to any resources. This also adds to the project attribute's knownAssetTags
```

File: Sources/XcodeGenKit/PBXProjGenerator.swift [hunk 2]
```diff
@@ -233,8 +233,15 @@ public class PBXProjGenerator {
             .sorted { $0.nameOrPath.localizedStandardCompare($1.nameOrPath) == .orderedAscending }
             .map { $0 }
 
+        let assetTags = Array(Set(project.targets
+            .map { target in
+                target.sources.map { $0.resourceTags }.flatMap { $0 }
+            }.flatMap { $0 }
+        ))
```
Comment (yonaskolb) [inline Sources/XcodeGenKit/PBXProjGenerator.swift]: Let's sort them as well, so it's deterministic. If you add a `sorted()` on the end, then the Array initializer isn't needed either, as sorted returns an array

File: Sources/XcodeGenKit/SourceGenerator.swift [hunk 3]
```diff
@@ -85,6 +85,7 @@ class SourceGenerator {
         let fileReference = fileReference ?? fileReferencesByPath[path.string.lowercased()]!
         var settings: [String: Any] = [:]
         var attributes: [String] = targetSource.attributes
+        let resourceTags: [String] = targetSource.resourceTags
```
Comment (yonaskolb) [inline Sources/XcodeGenKit/SourceGenerator.swift]: This `resourceTags` property is only used once, so let's just remove it and access `targetSource.resourceTags` inline down below

File: Sources/XcodeGenKit/PBXProjGenerator.swift [hunk 4]
```diff
@@ -233,8 +233,15 @@ public class PBXProjGenerator {
...
+        
         let projectAttributes: [String: Any] = ["LastUpgradeCheck": project.xcodeVersion]
             .merged(project.attributes)
+            .merged(["knownAssetTags" : assetTags])
```
Comment (yonaskolb) [inline Sources/XcodeGenKit/PBXProjGenerator.swift]: Only add this if `assetTags` is not empty, otherwise we'll cause a diff in the project

Comment (yonaskolb) [reply Sources/XcodeGenKit/PBXProjGenerator.swift]: (This is why the tests are failing by the way)

File: Sources/XcodeGenKit/SourceGenerator.swift [hunk 5]
```diff
@@ -124,6 +125,10 @@ class SourceGenerator {
         if !attributes.isEmpty {
             settings["ATTRIBUTES"] = attributes
         }
+        
+        if !resourceTags.isEmpty {
+            settings["ASSET_TAGS"] = resourceTags
```
Comment (yonaskolb) [inline Sources/XcodeGenKit/SourceGenerator.swift]: Let's also only add these if `chosenBuildPhase == .resources`

File: Sources/ProjectSpec/TargetSource.swift [hunk 6]
```diff
@@ -192,6 +195,7 @@ extension TargetSource: JSONObjectConvertible {
 
         createIntermediateGroups = jsonDictionary.json(atKeyPath: "createIntermediateGroups")
         attributes = jsonDictionary.json(atKeyPath: "attributes") ?? []
+        resourceTags = jsonDictionary.json(atKeyPath: "resourceTags") ?? []
```
Comment (yonaskolb) [inline Sources/ProjectSpec/TargetSource.swift]: You'll need to add a line in `toJSONValue` down below as well

Comment (sipao) [reply Sources/ProjectSpec/TargetSource.swift]: e038d9b

File: Sources/XcodeGenKit/PBXProjGenerator.swift (hunk 2 above)
Comment (sipao) [reply Sources/XcodeGenKit/PBXProjGenerator.swift]: 2c81e80

File: Sources/XcodeGenKit/PBXProjGenerator.swift (hunk 4 above)
Comment (sipao) [reply Sources/XcodeGenKit/PBXProjGenerator.swift]: bd567f6

File: Sources/XcodeGenKit/SourceGenerator.swift (hunk 3 above)
Comment (sipao) [reply Sources/XcodeGenKit/SourceGenerator.swift]: 91c45bf

File: Sources/XcodeGenKit/SourceGenerator.swift (hunk 5 above)
Comment (sipao) [reply Sources/XcodeGenKit/SourceGenerator.swift]: 213e922

**Comment by sipao** [comment]
Thanks for your review, @yonaskolb!

I fixed issues you reviewed.

**Pull request merged**